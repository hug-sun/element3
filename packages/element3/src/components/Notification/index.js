export { Notification } from './src/Notification.js'
