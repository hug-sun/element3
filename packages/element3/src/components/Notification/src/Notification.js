import { isVNode } from 'vue'
import notificationComponent from './Notification.vue'
import { createComponent } from '../../../../src/use/component.js'
import { PopupManager } from '../../../../src/utils/popup'

const instanceList = []
const INTERVAL_HEIGHT = 16

export function Notification(options) {
  return createNotification(mergeProps(options))
}

function addInstance(instance) {
  instanceList.push(instance)
}

function createNotification(options) {
  const instance = createNotificationByOpts(options)
  setZIndex(instance)
  addInstance(instance)
  addToBody(instance)
  return instance.proxy
}

function createNotificationByOpts(opts) {
  if (isVNode(opts.message)) {
    return createComponent(notificationComponent, opts, () => opts.message)
  }

  return createComponent(notificationComponent, opts)
}

function setZIndex(instance) {
  instance.vnode.el.style.zIndex = PopupManager.nextZIndex()
}

function addToBody(instance) {
  document.body.append(instance.vnode.el)
}

function mergeProps(options) {
  const position = options.position || 'top-right'

  const verticalOffset = calculateVerticalOffset(position)

  const defaultOptions = {
    position,
    verticalOffset
  }
  const userOnClose = options?.onClose
  delete options?.onClose
  defaultOptions.onClose = (instance) => {
    closeNotification(instance)
    if (userOnClose) userOnClose(instance.proxy)
  }

  const userOnClick = options?.onClick
  delete options?.onClick
  defaultOptions.onClick = (instance) => {
    if (userOnClick) userOnClick(instance.proxy)
  }

  if (typeof options === 'string' || isVNode(options)) {
    defaultOptions.message = options
    return defaultOptions
  }
  return Object.assign({}, defaultOptions, options)
}

function calculateVerticalOffset(position, offset = 0) {
  let verticalOffset = offset

  instanceList
    .filter((instance) => instance.props.position === position)
    .forEach((instance) => {
      verticalOffset += (instance.vnode.el.offsetHeight || 0) + INTERVAL_HEIGHT
    })
  verticalOffset += INTERVAL_HEIGHT

  return verticalOffset
}

function closeNotification(instance) {
  updatePosition(instance)
}

function updatePosition(closeInstance) {
  const currentInstanceIndex = getIndexByInstance(closeInstance)
  if (currentInstanceIndex < 0) return

  const instance = instanceList[currentInstanceIndex]
  const len = instanceList.length
  removeInstance(instance)
  if (len <= 1) return
  const position = instance.props.position
  const removedHeight = instance.vnode.el.offsetHeight
  for (let i = currentInstanceIndex; i < len - 1; i++) {
    if (instanceList[i].props.position === position) {
      instanceList[i].vnode.el.style[
        instance.props.position.startsWith('top') ? 'top' : 'bottom'
      ] =
        parseInt(
          instanceList[i].vnode.el.style[
            instance.props.position.startsWith('top') ? 'top' : 'bottom'
          ],
          10
        ) -
        removedHeight -
        INTERVAL_HEIGHT +
        'px'
    }
  }
}

function removeInstance(instance) {
  instanceList.splice(getIndexByInstance(instance), 1)
}

function getIndexByInstance(instance) {
  return instanceList.findIndex((i) => i.uid == instance.uid)
}

;['success', 'warning', 'info', 'error'].forEach((type) => {
  Notification[type] = (options) => {
    if (typeof options === 'string' || isVNode(options)) {
      options = {
        message: options
      }
    }
    options.type = type
    return Notification(options)
  }
})

Notification.closeAll = () => {
  instanceList.forEach((instance) => {
    instance.proxy.close()
    removeInstance(instance)
  })
}
