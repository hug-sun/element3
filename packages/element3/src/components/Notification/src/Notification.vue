<template>
  <transition
    name="el-notification-fade"
    @after-leave="handleAfterLeave"
    appear
  >
    <div
      :class="['el-notification', customClass, horizontalClass]"
      v-show="visible"
      :style="positionStyle"
      @click="handleClick"
      @keydown="handleKeydown"
      @mouseenter="handleMouseenter"
      @mouseleave="handleMouseleave"
      role="alert"
    >
      <i
        class="el-notification__icon"
        :class="typeClass || iconClass"
        v-if="type || iconClass"
      ></i>
      <div class="el-notification__group">
        <h2 class="el-notification__title" v-text="title"></h2>
        <div class="el-notification__content" v-show="message">
          <slot>
            <p v-if="!dangerouslyUseHTMLString">{{ message }}</p>
            <p v-else v-html="message"></p>
          </slot>
        </div>
        <div
          class="el-notification__closeBtn el-icon-close"
          v-if="showClose"
          @click="handleClose"
        ></div>
      </div>
    </div>
  </transition>
</template>

<script>
import { computed, getCurrentInstance, ref, defineComponent } from 'vue'
import { notificationProps } from './props.js'
export default defineComponent({
  name: 'ElNotification',
  props: notificationProps,
  emits: ['close'],
  setup(props, { emit }) {
    const instance = getCurrentInstance()
    const visible = ref(true)
    const verticalOffsetVal = ref(props.verticalOffset)

    const typeClass = computed(() => {
      return props.type ? `el-icon-${props.type}` : ''
    })

    const horizontalClass = computed(() => {
      return props.position.endsWith('right') ? 'right' : 'left'
    })

    const verticalProperty = computed(() => {
      return props.position.startsWith('top') ? 'top' : 'bottom'
    })

    const positionStyle = computed(() => {
      return {
        [verticalProperty.value]: `${verticalOffsetVal.value}px`
      }
    })

    function _click() {
      emit('click', instance)
    }
    function handleClick() {
      _click()
    }

    function click() {
      _click()
    }

    function _close() {
      clearTimeout(timer)
      emit('close', instance)
      visible.value = false
    }
    function handleClose() {
      _close()
    }

    function close() {
      _close()
    }

    let timer
    function delayClose() {
      if (props.duration > 0) {
        timer = setTimeout(() => {
          _close()
        }, props.duration)
      }
    }

    const handleKeydown = (e) => {
      if (e.keyCode === 46 || e.keyCode === 8) {
        clearTimeout(timer)
      } else if (e.keyCode === 27) {
        _close()
      } else {
        delayClose()
      }
    }

    function handleMouseenter() {
      clearTimeout(timer)
    }
    function handleMouseleave() {
      delayClose()
    }

    function handleAfterLeave() {
      instance.vnode.el.parentElement?.removeChild(instance.vnode.el)
    }

    delayClose()

    return {
      close,
      click,
      visible,
      typeClass,
      positionStyle,
      horizontalClass,
      verticalProperty,
      verticalOffsetVal,
      handleClose,
      handleClick,
      handleKeydown,
      handleMouseenter,
      handleMouseleave,
      handleAfterLeave
    }
  }
})
</script>
