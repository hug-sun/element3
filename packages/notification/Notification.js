import { defineComponent, createVNode, render, isVNode } from 'vue'
import Main from './Notification.vue'
import merge from '../../src/utils/merge'
import { PopupManager } from '../../src/utils/popup'

const NotificationConstructor = defineComponent(Main)

let instance
const instances = []
let seed = 1

const Notification = function (options) {
  // check ssr
  if (window === undefined) return

  const userOnClose = options.onClose
  const id = 'notification_' + seed++
  const position = options.position || 'top-right'
  let verticalOffset = options.offset || 0
  instances
    .filter((instance) => instance.component.props.position === position)
    .forEach((instance) => {
      verticalOffset += (instance.el.offsetHeight || 0) + 16
    })
  verticalOffset += 16
  const nextZIndex = PopupManager.nextZIndex()
  options = merge(
    {
      id,
      position,
      verticalOffset,
      zIndex: nextZIndex,
      onClose() {
        Notification.close(id, userOnClose)
      }
    },
    options
  )
  instance = createVNode(
    NotificationConstructor,
    options,
    isVNode(options.message)
      ? {
          default: () => options.message
        }
      : null
  )

  // 防止render直接到body上导致覆盖原来的
  const container = document.createElement('div')

  render(instance, container)
  instance.el.style.zIndex = nextZIndex
  instances.push(instance)
  document.body.appendChild(instance.el)
  return {
    close() {
      Notification.close(id, userOnClose)
    }
  }
}

;['success', 'warning', 'info', 'error'].forEach((type) => {
  Notification[type] = (options) => {
    if (typeof options === 'string' || isVNode(options)) {
      options = {
        message: options
      }
    }
    options.type = type
    return Notification(options)
  }
})

Notification.close = function (id, userOnClose) {
  const index = instances.findIndex((instance) => {
    return instance.component.props.id === id
  })
  if (index === -1) {
    return
  }

  const instance = instances[index]

  if (typeof userOnClose === 'function') {
    userOnClose(instance)
  }

  const len = instances.length
  instances.splice(index, 1)

  if (len <= 1) return

  const position = instance.component.props.position
  const removedHeight = instance.el.offsetHeight
  for (let i = index; i < len - 1; i++) {
    if (instances[i].component.props.position === position) {
      instances[i].el.style[
        instance.component.props.position.startsWith('top') ? 'top' : 'bottom'
      ] =
        parseInt(
          instances[i].el.style[
            instance.component.props.position.startsWith('top')
              ? 'top'
              : 'bottom'
          ],
          10
        ) -
        removedHeight -
        16 +
        'px'
    }
  }
}

Notification.closeAll = function () {
  for (let i = instances.length - 1; i >= 0; i--) {
    instances[i].component.ctx.close()
  }
}

export default Notification
