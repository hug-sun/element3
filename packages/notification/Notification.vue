<template>
  <transition name="el-notification-fade" appear>
    <div
      :class="['el-notification', customClass, horizontalClass]"
      v-show="visible"
      :style="positionStyle"
      @mouseenter="clearTimer()"
      @mouseleave="startTimer()"
      @click="click"
      role="alert"
    >
      <i
        class="el-notification__icon"
        :class="[typeClass, iconClass]"
        v-if="type || iconClass"
      ></i>
      <div
        class="el-notification__group"
        :class="{ 'is-with-icon': typeClass || iconClass }"
      >
        <h2 class="el-notification__title" v-text="title"></h2>
        <div class="el-notification__content" v-show="message">
          <slot>
            <p v-if="!dangerouslyUseHTMLString">{{ message }}</p>
            <p v-else v-html="message"></p>
          </slot>
        </div>
        <div
          class="el-notification__closeBtn el-icon-close"
          v-if="showClose"
          @click.stop="close"
        ></div>
      </div>
    </div>
  </transition>
</template>

<script>
import {
  computed,
  getCurrentInstance,
  onBeforeUnmount,
  onMounted,
  ref,
  watch
} from 'vue'
const typeMap = {
  success: 'success',
  info: 'info',
  warning: 'warning',
  error: 'error'
}
export default {
  name: 'ElNotification',
  props: {
    customClass: { type: String, default: '' },
    dangerouslyUseHTMLString: { type: Boolean, default: false },
    duration: { type: Number, default: 4500 },
    iconClass: { type: String, default: '' },
    id: { type: String, default: '' },
    verticalOffset: { type: Number, default: 0 },
    message: [String, Object],
    position: {
      type: String,
      default: 'top-right'
    },
    onClose: null,
    onClick: null,
    showClose: { type: Boolean, default: true },
    title: { type: String, default: '' },
    type: { type: String, default: '' }
  },
  emits: ['close'],
  setup(props) {
    // eslint-disable-next-line vue/no-setup-props-destructure
    const { duration, onClose, onClick } = props
    const typeClass = computed(() => {
      return props.type && typeMap[props.type]
        ? `el-icon-${typeMap[props.type]}`
        : ''
    })
    const horizontalClass = computed(() => {
      return props.position.indexOf('right') > -1 ? 'right' : 'left'
    })
    const verticalProperty = computed(() => {
      return props.position.startsWith('top') ? 'top' : 'bottom'
    })
    const positionStyle = computed(() => {
      return {
        [verticalProperty.value]: `${props.verticalOffset}px`
        // zIndex,
      }
    })
    const instance = getCurrentInstance()
    const visible = ref(true)
    const closed = ref(false)
    const timer = ref(0)
    const destroyElement = () => {
      instance.ctx.$el.parentNode && instance.ctx.$el.parentNode.removeChild(instance.ctx.$el)
    }
    const clearTimer = () => {
      clearTimeout(timer.value)
    }
    const close = () => {
      closed.value = true
      if (typeof onClose === 'function') {
        onClose()
      }
    }
    const startTimer = () => {
      if (duration > 0) {
        timer.value = setTimeout(() => {
          if (!closed.value) {
            close()
          }
        }, duration)
      }
    }
    const keydown = (e) => {
      if (e.keyCode === 46 || e.keyCode === 8) {
        clearTimer() // detele 取消倒计时
      } else if (e.keyCode === 27) {
        // esc关闭消息
        if (!closed.value) {
          close()
        }
      } else {
        startTimer() // 恢复倒计时
      }
    }
    const click = () => {
      if (typeof onClick === 'function') {
        onClick()
      }
    }
    watch(closed, (newVal) => {
      if (newVal) {
        visible.value = false
        instance.ctx.$el.addEventListener('transitionend', destroyElement)
      }
    })
    onMounted(() => {
      if (duration > 0) {
        timer.value = setTimeout(() => {
          if (!closed.value) {
            close()
          }
        }, duration)
      }
      visible.value = true
      document.addEventListener('keydown', keydown)
    })
    onBeforeUnmount(() => {
      document.removeEventListener('keydown', keydown)
    })
    return {
      typeClass,
      horizontalClass,
      verticalProperty,
      positionStyle,
      visible,
      closed,
      timer,
      click,
      clearTimer,
      startTimer,
      close
    }
  }
}
</script>
