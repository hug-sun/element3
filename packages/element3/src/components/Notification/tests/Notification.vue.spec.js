import { mount, flushPromises } from '@vue/test-utils'
import Notification from '../src/Notification.vue'
import { h } from 'vue'

describe('Notification.vue', () => {
  test('snapshot', () => {
    const wrapper = mount(Notification)
    expect(wrapper.element).toMatchSnapshot()
  })

  it('title', () => {
    const title = 'this is a title'
    const wrapper = mount(Notification, {
      props: {
        title
      }
    })
    expect(wrapper.get('.el-notification__title').text()).toContain(title)
  })

  it('the message type is string', () => {
    const message = 'this is a message'
    const wrapper = mount(Notification, {
      props: {
        message
      }
    })
    expect(wrapper.get('.el-notification__content').text()).toContain(message)
  })

  it('the message type is VNode', () => {
    const message = h('p', 'foo')
    const wrapper = mount(Notification, {
      slots: {
        default: message
      }
    })
    expect(wrapper.get('.el-notification__content').get('p').html()).toContain(
      '<p>foo</p>'
    )
  })

  it('dangerouslyUseHTMLString', () => {
    const wrapper = mount(Notification, {
      props: {
        dangerouslyUseHTMLString: true,
        message: "<div id='foo'>foo</div>"
      }
    })
    expect(wrapper.find('#foo').exists()).toBe(true)
  })

  describe('type', () => {
    test('show type icon', () => {
      const type = 'info'
      const wrapper = mount(Notification, {
        props: {
          type
        }
      })

      expect(wrapper.find(`.el-icon-${type}`).exists()).toBe(true)
      expect(wrapper.find('.el-notification__icon').exists()).toBe(true)
    })
  })

  it('iconClass', () => {
    const iconClass = 'el-icon-setting'
    const wrapper = mount(Notification, {
      props: {
        iconClass
      }
    })
    expect(wrapper.find(`.${iconClass}`).exists()).toBe(true)
    expect(wrapper.find('.el-notification__icon').exists()).toBe(true)
  })

  it('iconClass should be overridden', () => {
    const type = 'info'
    const iconClass = 'el-icon-setting'
    const wrapper = mount(Notification, {
      props: {
        type,
        iconClass
      }
    })
    expect(wrapper.find(`.el-icon-${type}`).exists()).toBe(true)
    expect(wrapper.find(`.${iconClass}`).exists()).toBe(false)
    expect(wrapper.find('.el-notification__icon').exists()).toBe(true)
  })

  it('customize the name of the class', () => {
    const customClass = 'foo'
    const wrapper = mount(Notification, {
      props: {
        customClass
      }
    })
    expect(wrapper.get('.el-notification').classes()).toContain(customClass)
  })

  it('position', () => {
    const position = 'bottom-right'
    const wrapper = mount(Notification, {
      props: {
        position
      }
    })
    expect(wrapper.find('.el-notification').classes()).toContain('right')
    expect(wrapper.vm.verticalProperty).toBe('bottom')
    expect(wrapper.find('.el-notification').element.style.bottom).toBe('0px')
  })

  it('verticalOffset', () => {
    const verticalOffset = 50
    const wrapper = mount(Notification, {
      props: {
        verticalOffset
      }
    })
    expect(wrapper.vm.verticalProperty).toBe('top')
    expect(wrapper.find('.el-notification').element.style.top).toBe(
      `${verticalOffset}px`
    )
  })

  it('set both position and verticalOffset', () => {
    const position = 'bottom-right'
    const verticalOffset = 50
    const wrapper = mount(Notification, {
      props: {
        position,
        verticalOffset
      }
    })
    expect(wrapper.vm.verticalProperty).toBe('bottom')
    expect(wrapper.find('.el-notification').element.style.bottom).toBe(
      `${verticalOffset}px`
    )
  })

  it('update verticalOffset', async () => {
    const wrapper = mount(Notification, {
      props: {
        verticalOffset: 50
      }
    })
    expect(wrapper.find('.el-notification').element.style.top).toBe('50px')
    wrapper.vm.verticalOffsetVal = 100
    await flushPromises()
    expect(wrapper.find('.el-notification').element.style.top).toBe('100px')
  })

  describe('close', () => {
    it('set the showClose ', () => {
      const showClose = true
      const wrapper = mount(Notification, {
        props: {
          showClose
        }
      })
      expect(wrapper.find('.el-notification__closeBtn').exists()).toBe(true)
      expect(wrapper.find('.el-icon-close').exists()).toBe(true)
    })

    it('click close btn', async () => {
      const showClose = true
      const wrapper = mount(Notification, {
        props: {
          showClose
        }
      })
      const closeBtn = wrapper.get('.el-notification__closeBtn')
      await closeBtn.trigger('click')
      expect(wrapper.get('.el-notification').isVisible()).toBe(false)
    })

    it('set the duration to 1000, notification is closed when in 1000ms', async () => {
      jest.useFakeTimers()

      const wrapper = mount(Notification, {
        props: {
          duration: 1000
        }
      })
      jest.runTimersToTime(1000)
      await flushPromises()
      expect(wrapper.get('.el-notification').isVisible()).toBe(false)
    })

    it('set the duration to 0, the Notification does not turn off ', async () => {
      jest.useFakeTimers()
      const wrapper = mount(Notification, {
        props: {
          duration: 0
        }
      })
      jest.runTimersToTime(5000)
      await flushPromises()
      expect(wrapper.get('.el-notification').isVisible()).toBe(true)
    })

    it('should called onClose', () => {
      let proxy
      const onClose = (componentInstance) => {
        proxy = componentInstance.proxy
      }
      const wrapper = mount(Notification, {
        props: {
          onClose,
          showClose: true
        }
      })
      wrapper.get('.el-notification__closeBtn').trigger('click')
      expect(proxy.close).toBeTruthy()
    })
  })

  describe('notification onClick', () => {
    it('should called onClick', () => {
      let proxy
      const onClick = (componentInstance) => {
        proxy = componentInstance.proxy
      }
      const wrapper = mount(Notification, {
        props: {
          onClick,
          showClose: true
        }
      })
      wrapper.get('.el-notification').trigger('click')
      expect(proxy.click).toBeTruthy()
    })
  })

  it('should close notification when call close function in the componentInstance', async () => {
    const wrapper = mount(Notification)

    wrapper.vm.close()
    await flushPromises()
    expect(wrapper.get('.el-notification').isVisible()).toBe(false)
  })

  it('should clear timeout', () => {
    jest.useFakeTimers()

    const wrapper = mount(Notification, {
      props: {
        duration: 1000,
        showClose: true
      }
    })

    wrapper.get('.el-notification__closeBtn').trigger('click')
    jest.runTimersToTime(1000)
    expect(wrapper.emitted('close')).toBeTruthy()
    expect(wrapper.emitted('close')[0][0].proxy.close).toBeTruthy()
  })

  it('should not closed when mouseenter', async () => {
    jest.useFakeTimers()

    const wrapper = mount(Notification, {
      props: {
        duration: 1000
      }
    })

    wrapper.get('.el-notification').trigger('mouseenter')
    jest.runTimersToTime(1000)
    expect(wrapper.emitted('close')).toBeFalsy()
  })

  it('should closed when mouseleave', async () => {
    jest.useFakeTimers()

    const wrapper = mount(Notification, {
      props: {
        duration: 1000
      }
    })

    wrapper.get('.el-notification').trigger('mouseenter')
    wrapper.get('.el-notification').trigger('mouseleave')
    jest.runTimersToTime(1000)
    expect(wrapper.emitted('close')).toBeTruthy()
  })

  it('keyCode is 46', async () => {
    jest.useFakeTimers()
    const wrapper = mount(Notification)
    wrapper.get('.el-notification').trigger('keydown', { keyCode: 46 })

    jest.runTimersToTime(5000)
    await flushPromises()
    expect(wrapper.get('.el-notification').isVisible()).toBe(true)
  })
  it('keyCode is 27', async () => {
    let proxy
    const onClose = (componentInstance) => {
      proxy = componentInstance.proxy
    }
    const wrapper = mount(Notification, {
      props: {
        onClose
      }
    })
    wrapper.get('.el-notification').trigger('keydown', { keyCode: 27 })
    expect(proxy.close).toBeTruthy()
  })

  it('keyCode is 100', async () => {
    jest.useFakeTimers()

    const wrapper = mount(Notification)
    jest.runTimersToTime(4500)
    await flushPromises()

    wrapper.get('.el-notification').trigger('keydown', { keyCode: 100 })
    expect(wrapper.get('.el-notification').isVisible()).toBe(false)
  })
})
