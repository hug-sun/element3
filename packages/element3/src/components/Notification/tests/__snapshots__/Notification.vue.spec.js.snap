// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification.vue snapshot 1`] = `
<transition-stub>
  <div
    class="el-notification  right"
    role="alert"
    style="top: 0px;"
  >
    <!--v-if-->
    <div
      class="el-notification__group"
    >
      <h2
        class="el-notification__title"
      />
      <div
        class="el-notification__content"
        style="display: none;"
      >
        
        <p />
        
      </div>
      <div
        class="el-notification__closeBtn el-icon-close"
      />
    </div>
  </div>
</transition-stub>
`;
